features/device-config/monitor/src/main/java/org/opennms/features/deviceconfig/monitors/DeviceConfigMonitor.java

```
@@ -0,0 +1,151 @@
+        long ttl = ttlValue != null ? ttlValue : DEFAULT_TTL;
+        Optional<SshScriptingService.Failure> sshResult =
+                sshScriptingService.execute(script, user, password, svc.getIpAddr(), port, new HashMap<>(), duration);
+        CompletableFuture<byte[]> configObj = tftpReceiver.getConfigFuture();
+        try {
+            byte[] config = configObj.get(ttl, TimeUnit.MILLISECONDS);
+            PollStatus pollStatus = PollStatus.up();
+            pollStatus.setDeviceConfig(config);
+            return pollStatus;
+        } catch (TimeoutException | InterruptedException | ExecutionException e) {
+            LOG.error("Config retrieval failed", e);
+            return PollStatus.unresponsive("Config retrieval failed : " + e.getMessage());
+        }
+    }
+
+    static private class TftpReceiver implements TftpFileReceiver {
+
+        private final CompletableFuture<byte[]> configFuture = new CompletableFuture<>();
+        private final InetAddress serviceAddress;
+
+        public TftpReceiver(InetAddress serviceAddress) {
+            this.serviceAddress = serviceAddress;
+        }
+
+        @Override
+        public void onFileReceived(InetAddress address, String fileName, byte[] content) {
+            if (serviceAddress.equals(address)) {
+                configFuture.complete(content);
+            }
+        }
+
+        public CompletableFuture<byte[]> getConfigFuture() {
+            return configFuture;
+        }
+    }
+
+    public void setTftpServer(TftpServer tftpServer) {
+        this.tftpServer = tftpServer;
+    }
+
+    public void setSshScriptingService(SshScriptingService sshScriptingService) {
+        this.sshScriptingService = sshScriptingService;
+    }
+
+    private String getObjectAsStringFromParams(Map<String, Object> params, String key) {
+        Object obj = params.get(key);
+        if (obj instanceof String) {
+            return (String) obj;
+        }
+        return null;
```

Maybe throw an exception if a non-string value was configured.